# BMS500C
Tutorials for BMS500C 2023 class

The overall objectives of this session are 
  1) to become familiar with navigating in a command-line (Linux) environment.
  2) to perform basic and more sophisticated command-line operations (such as using regex, piping, and for-loops) to parse and manipuate data
  3) to learn the basic operations for performing a reference-based assembly, calling variants, and evaluating the results
  4) to generate viral phylogenies for assess relationships among isolates and looking for temporal and geographical signals.

West Nile Virus 
We will download next-generation sequencing (NGS) data from the Sequence Read Archive (SRA) on NCBI for several West Nile Virus isolates.
After mapping the sequence reads to a reference genome, we will generate some summary statistics to evaluate the quality of our assemblies and 
perform variant analyses to idenfity high-quality single-nucleotide polymorphisms (SNPs).  This information will be used to generate consensus
genomes for our isolates.  We will infer their relationships to other contextual WNV genomes from North America by generating a phylogeny and evaluate
whether there is a geographical and/or temporal signal to the data.
